linera-client/src/chain_listener.rs

```diff
@@ -122,141 +137,58 @@ impl<C: ClientContext> ChainListener<C> {
-            chain_ids
+            let chain_ids = guard.wallet().chain_ids().into_iter();
+            chain_ids.chain(iter::once(guard.wallet().genesis_admin_chain()))
         };
         for chain_id in chain_ids {
-            self.run_with_chain_id(chain_id);
-        }
-    }
-
-    /// Spawns a task running the listener for the given chain, if it is not already running.
-    #[instrument(level = "trace", skip_all, fields(?chain_id))]
-    fn run_with_chain_id(&self, chain_id: ChainId) {
-        if !self.listening.lock().unwrap().insert(chain_id) {
-            // If we are already listening to notifications, there's nothing to do.
-            // This can happen if we download a child before the parent
-            // chain, and then process the OpenChain message in the parent.
-            return;
-        }
-        let listener = self.clone();
-        let _handle = linera_base::task::spawn(
-            async move {
-                if let Err(err) = listener.run_client_stream(chain_id).await {
-                    error!("Stream for chain {} failed: {}", chain_id, err);
-                }
-            }
-            .in_current_span(),
-        );
-    }
-
-    #[instrument(level = "trace", skip_all, fields(?chain_id))]
-    async fn run_client_stream(self, chain_id: ChainId) -> Result<(), Error> {
-        let client = self.context.lock().await.make_chain_client(chain_id)?;
-        let (event_tx, event_rx) = mpsc::channel(1);
-        if client.admin_id() != chain_id {
-            let mut admin_event_tx = event_tx.clone();
-            let mut admin_stream = client.subscribe_to(client.admin_id()).await?;
-            let _handle = linera_base::task::spawn(
-                async move {
-                    while let Some(notification) = admin_stream.next().await {
-                        if let Reason::NewBlock { .. } = notification.reason {
-                            if admin_event_tx.send(()).await.is_err() {
-                                return; // The receiver was dropped.
-                            }
-                        }
-                    }
-                }
-                .in_current_span(),
-            );
-        }
-        let (listener, _listen_handle, local_stream) = client.listen().await?;
-        client.synchronize_from_validators().await?;
-        let _handle = linera_base::task::spawn(listener.in_current_span());
-        let client_listener = ChainClientListener::new(client, self);
-        client_listener.run(local_stream, event_rx).await
-    }
-}
-
-impl<C: ClientContext> ChainClientListener<C> {
-    /// Creates a new chain client listener.
-    fn new(client: ContextChainClient<C>, listener: ChainListener<C>) -> Self {
-        Self {
-            client,
-            listener,
-            timeout: Timestamp::from(u64::MAX),
+            self.listen(chain_id).await?;
         }
+        let _handler = linera_base::task::spawn(self.run_notification_loop().in_current_span());
```

Comment:
I'm not totally clear what this spawn is for.  Can we not let the caller decide how to poll this chain listener?

Reply:
Done in https://github.com/linera-io/linera-protocol/pull/3745/commits/ba3a9fad87b93ac1c44adf5c95c182ec186b0ef7.